File: backend/models/schemas.py
from pydantic import BaseModel, field_serializer, field_validator, Field, SerializationInfo
from typing import Union, Optional, List
from datetime import datetime
from bson import ObjectId
from beanie import PydanticObjectId
import pydantic


class UserDTO(BaseModel):
    id: str 
    username: str
    email: str
    created_at: datetime


class ConversationDTO(BaseModel):
    id: str 
    user_id: str
    session_id: str
    name: str
    started_at: datetime
    last_active: datetime
    is_archived: bool
    flags: dict

    
class ChatMessageDTO(BaseModel):
    id: str
    user_id: str
    conversation_id: str
    session_id: str
    timestamp: datetime
    role: str
    content: str
    file_id: Optional[str] | None = None
   

class PromptSettingsDTO(BaseModel):
    id: str | None = None
    user_id: str
    display_name: str | None = None
    custom_prompt: str | None = None
    occupation: str | None = None
    interests: str | None = None
    about_me: str | None = None
    updated_at: str | None = None # ISO string format with milliseconds precision


class FileDoc(BaseModel):
    file_id: str
    file_type: str
    file_size: int
    file_name: str 
    file_url: str




    
  

File: my-app/src/app/types/schemas.ts
import { z } from "zod";

export const MessageSchema = z.object({
    id: z.string(),
    user_id: z.string(),
    session_id: z.string(),
    conversation_id: z.string(),
    role: z.string(),
    content: z.string(),
    timestamp: z.coerce.date().optional(),
    flags: z.any(),
    file_id: z.string().nullable(),
})

export type Message = z.infer<typeof MessageSchema>;


export const ConversationSchema = z.object({
    id: z.string(),
    user_id: z.string(),
    session_id: z.string(),
    name: z.string(),
    started_at: z.coerce.date().optional(),
    last_active: z.coerce.date().optional(),
    is_archived: z.boolean(),
    flags: z.any(),
})

export type Conversation = z.infer<typeof ConversationSchema>;

export const UserSchema = z.object({
    id: z.string(), // Pydantic ObjectId serialized to string
    username: z.string(),
    email: z.string().email(),
    created_at: z.coerce.date().optional() // assuming ISO string from backend
  });
  
  export type User = z.infer<typeof UserSchema>;


export const PromptSettingsSchema = z.object({
    id: z.string().optional(),
    user_id: z.string(),
    display_name: z.string(),
    custom_prompt: z.string(),
    occupation: z.string(),
    interests: z.string(),
    about_me: z.string(),
    updated_at: z.coerce.date().optional()// assuming ISO string from backend
  });
  
  export type Promptsettings = z.infer<typeof PromptSettingsSchema>;

export const FileAttachmentSchema = z.object({
    file_id: z.string(),
    file_type: z.string().optional(),
    file_size: z.number().optional(),
    file_name: z.string().optional(),
    file_url: z.string(),
}).transform((data) =>({
    id: data.file_id,
    type: data.file_type,
    size: data.file_size,
    name: data.file_name,
    url: data.file_url
}));

export type FileAttachment = z.infer<typeof FileAttachmentSchema>;
